### DISCUSSION TITLE
gatsby-source-shopify: Version 4 to 5 migration

### DISCUSSION THREAD

**Discussion by wildpow** [gatsbyjs/gatsby#32093] [Help]
I thought it would be good to have a place where people can share their experiences and useful information while testing out V5 release candidates.

**Comment by wildpow** [comment]
### Issue 1 - `priceRange` is now `priceRangeV2`
I believe both fields existed in V4. Version 5 only exposes `priceRangeV2`

**Comment by wildpow** [comment]
### Issue 2 - `shopifyId` is now ` storefrontId`
Well, not exactly, In version 4, querying `shopifyId` would produce something like this `Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0LzY2NTY2Mjg4ODM2NDA`. Because the new plug-in uses the bulk operations API, the same field produces something like this `gid://shopify/Product/4525916422209`, which in its self not that big of a deal unless you are using these IDs to combine data from multiple sources or something else exotic. The field `storefrontId` retains the id that the `shopifyId` did in version 4.

**Reply by Hahlh** [reply]
Hey @wildpow!

Just wanted to let you know realy quick that this was really useful info and spared me lots of pain debugging. 

Thank you for sharing!

**Comment by wildpow** [comment]
### Issue 2.5 - `shopifyId` is now `storefronId`
My store did not like the new `shopifyId` data.
I'm pretty sure this problem is in how I implemented my store and not with how `shopify-buy` handles ids. Instead of a rewrite, I just aliased `storefronId` with `shopifyId` in my queries like this:
```graphql
 variants {
          compareAtPrice
          price
          title
          shopifyId   // <------------------Old way
          shopifyId: storefrontId   // <---- New way
        }
```

**Comment by TrevHeath** [comment]
Noticed that `availableForSale` is not available on the schema.  Any idea what replaced it?  Currently `availableForSale`, from my understanding, considers both stock available and the continued overselling functionality.

**Comment by unimprobable** [comment]
Does v5 source Shopify blogs? In lower versions, there are allShopifyArticle and allShopifyBlog nodes.

**Comment by sslotsky** [comment]
@TrevHeath `availableForSale` is not available in the admin API, which is where we're sourcing from. Although I would think it's better to check this availability against the Storefront API on the front end since it seems like this could change quite between build times.

**Reply by dungle-scrubs** [reply]
@TrevHeath & @sslotsky 

I believe `availableForSale` is only accessible via `ProductVariant` in the Admin API, while in Storefront API it's accessible for both `Product` and `ProductVariant`

**Comment by sslotsky** [comment]
@unimprobable the admin API does not expose this content so we don't currently plan on sourcing it.

**Comment by unimprobable** [comment]
@sslotsky Thanks for the reply. That's disappointing as this is a breaking change for my project! Do you have any ideas how to source the Shopify Blog while using v5.0? I see it is exposed in the REST Admin API (not much help there I know).

**Comment by marcobiedermann** [comment]
I've noticed that `allShopifyCollection`, `shopifyShop`, `shopifyShopPolicy`, `allShopifyArticle` and so on have been removed. I've set read-access to all of them in the admin panel. Is this by intention or am I missing something?

**Comment by unimprobable** [comment]
> I've noticed that `allShopifyCollection`, `shopifyShop`, `shopifyShopPolicy`, `allShopifyArticle` and so on have been removed. I've set read-access to all of them in the admin panel. Is this by intention or am I missing something?

@marcobiedermann You can access the allShopifyCollection at least by specifying "collections" in the gatsby-config.js file.

`        shopifyConnections: ["collections"],
`

**Comment by sslotsky** [comment]
@unimprobable we use the bulk operations portion of the admin API to source everything. That said, I don't think I knew that they were available from the REST Admin API until now. We'd have to do some research to see if this resource type would support incremental builds, but if it does, I think we could reconsider it.

In the meantime you could potentially use `gatsby-source-graphql` to fetch this stuff from the Storefront API.

**Comment by unimprobable** [comment]
> In the meantime you could potentially use `gatsby-source-graphql` to fetch this stuff from the Storefront API.

Good idea, I will try that. Thanks, @sslotsky !

**Comment by sslotsky** [comment]
@marcobiedermann there are some things that aren't exposed by the admin/bulk API, and since that's how we currently source all of our data for v5, we share its limitations. That said, we'd love to understand what your use case is for any piece of data you need that we aren't currently exposing for you. So if you really need e.g. `shopifyShop` and `shopifyShopPolicy`, please open a new issue and we'll see if there's anything can do.

As for the others: `allShopifyArticle` is also what @unimprobable was asking about so feel free to reference our discussion there, and `allShopifyCollection` is available on an opt-in basis as pointed out by @unimprobable. It's quite possible that we should have included collections by default instead of opt-in.

**Comment by marcobiedermann** [comment]
@sslotsky 
Thank you very much for your quick reply.
Actually, I am using all sources, I've mentioned above.
I've set up a demo shop right here: https://marcobiedermann-gatsby-shopify.netlify.app/ (source: https://github.com/marcobiedermann/gatsby-shopify)
Of course, this is all fake data and nothing to use on a production application.

These are my use cases:
shopifyShop: Store name, currency
shopifyShopPolicy: Privacy, Refund, Terms of Services
allShopifyArticle: Blogs and Articles per blog and comments (note that you can have multiple blogs)
allShopifyCollection: I think collections and vendors are kinda interchangable

**Comment by unimprobable** [comment]
I am trying to use the downloadImages: true flag, but am getting timed out at the Shopify side (I think).

I used to solve this issue with a small paginationSize. Is there any way to affect the request like that now?

**Comment by sslotsky** [comment]
@marcobiedermann can you please open another issue as I requested previously? We will make this thread too noisy by discussing this here. Similarly @unimprobable there is already an issue open for this problem. 

In general let's try to keep this thread dedicated to collecting tips for the migration and not reporting new issues, please.

**Comment by rowanflack** [comment]
Another difference between plugin versions: 

variants.image.id = gid://shopify/ProductImage/22817961902275

**Comment by petertait** [comment]
Another difference seems to be the lack of custom domains for the checkout webUrl. From previous plugin:

```
// If you are running your shop on a custom domain, you need to use that
// as the shop name, without a trailing slash, for example:
// shopName: "gatsby-shop.com",
```

This is no longer mentioned, which results in the checkout domain being: `shopname.myshopify.com/...` rather than `shopname.com/...`

**Reply by unimprobable** [reply]
@petertait Did you ever figure out how to use a custom domain? I see people use subdomains like us.mycustomdomain.com instead of mycustomdomain.myshopify.com but I can't figure out how to do this with this plugin.

I can use postman etc to hit the Shopify API successfully with the subdomain as the endpoint, I just can't get it to work with gatsby-source-shopify. Wonder if they are all rolling their own storefronts.

**Reply by zamson** [reply]
Also need this for my store. Why was the feature removed from the plugin?

**Reply by dtcbyron** [reply]
For anyone still wondering about this, after speaking to the Shopify Plus team I learned that the checkout webUrl is determined by the URL you give to the storefront API. Thus for this plugin, whatever URL you provide to the shopName field will become the base URL of the checkout.webUrl field!

**Reply by unimprobable** [reply]
@byronlanehill Have you got that to work? I have not been able to. Thanks.

**Comment by smastrom-dolly** [comment]
If anyone is looking for `quantityAvailable`, it is now `inventoryQuantity`

**Comment by sahandnayebaziz** [comment]
Is this an official Shopify Gatsby plugin? If so, I am little surprised by the V4 → V5 transition without a guide or anything like that. If not, totally makes sense and of course I appreciate this great open source work. It's just a little bit surprising because going with something as large as Shopify, I would expect for renamed variables to all be listed out and not to have to hunt them down via what others have posted

**Comment by chang-ryan** [comment]
i'm fairly certain that this is not working as intended:

https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-shopify/src/query-builders/products-query.ts#L5-L7

the `salesChannel` query filter is not working

**Comment by wildpow** [comment]
Is there an option to set your default schema version for this plug-in in the `gatsby-config.js`? I feel like there was in v4 or am I crazy? Either something is down or there were breaking changes in the newest version of Shopify's schema because I have a project that was building just fine last week and now builds are failing. Just a theory since we haven't made any changes on our side.

**Comment by chrisvince** [comment]
It looks like there's no `priceV2`, only `price` on variants (which contains no currency).

**Reply by CalebBarnes** [reply]
This is needed.

**Reply by dungle-scrubs** [reply]
I'm noticing something strange regarding this.  When I start up a development server and visit `http://localhost:8000/___graphql` **before** visiting `http://localhost:8000/`, product variants do not appear to have `priceV2` in GraphiQL, as described above.

However, my site definitely uses `priceV2`.  As soon as I visit `http://localhost:8000/` in the browser, after refreshing GraphiQL, `priceV2` shows up.  

The site has never broken, so it looks like the only issue is that it's not appearing in GraphiQL initially.

**Comment by chrisvince** [comment]
I also just noticed there's no `image` on variants.

**Comment by jarodpeachey** [comment]
Hey all, my build is failing using v5 of this plugin. Can't figure out ANYTHING; I've set the API permissions to read access for everything on Shopify

Keep getting this error: `Received error 404 from Shopify: {"errors":"Not Found"}`

**Comment by presto2116** [comment]
Hey all, I have been able to make this migration to the new source plugin. One strange bug I am encountering is if i set `downloadImages: true` my images array does not follow the correct order which is causing issues with design/business. our galleries are set up in a very specific order and when `downloadImages: false`, the order is kept, but `downloadImages: true` i am guessing the order is created by whenever the image processing is completed. any work arounds or updates to fix this?

**Reply by dtcbyron** [reply]
Hey @presto2116, I am currently working with the Shopify team to create a new version of this plugin that solves this issue and many others. We are hoping to have it in production within the next few weeks, but in the meantime you can find the canary at https://www.npmjs.com/package/gatsby-source-shopify/v/6.2.0-alpha-atomic-shopify-source.16

Note that the canary is a WIP and is not production-ready as of Dec. 20 2021

**Reply by AlexanderProd** [reply]
@byronlanehill Is there any progress on this issue? I'm using the latest version of `gatsby-source-shopify` (6.5.2) but the problem with the incorrect order of images still exists.

**Reply by dtcbyron** [reply]
@AlexanderProd https://www.npmjs.com/package/gatsby-source-shopify/v/6.2.0-alpha-atomic-shopify-source.25 The Gatsby team and I have released a Canary for v7 of gatsby-source-shopify. It includes a bunch of breaking changes and updates, but fixed many of the issues that devs were experiencing, including the issue that you are facing. Give it a try, and if you have any questions feel free to reach out!

**Reply by AlexanderProd** [reply]
@byronlanehill 
> https://www.npmjs.com/package/gatsby-source-shopify/v/6.2.0-alpha-atomic-shopify-source.25 

When using that version I'm getting the following error.

```
GraphQLError: Cannot query field \"images\" on type \"ShopifyProduct\". Did you mean \"tags\"?
```

**Reply by dtcbyron** [reply]
@AlexanderProd I am not in the office today, but I will send what we have whipped up so far for a migration guide when I am. The biggest change is that the images field no longer exists and has been replaced with media. Media is an interface that allows querying for videos, images and 3D renderings. If you take a look at the GraphiQL explorer you can craft a query from there

**Reply by AlexanderProd** [reply]
@byronlanehill any update on this? :)

**Reply by mud1** [reply]
Hi, I have the exact same issue, image order is important for my shop. Any update on v7? Thank you.

**Comment by andrewwithjoy** [comment]
Hello all,

I'm in the progress of moving from Gatsby v3 to v4, and trying to bring gatsby-source-shopify along for the ride.  The configuration has changed from plugin v4 to v5, accessToken to storeUrl + password.  I'm very new to the Gatsby and Shopify ecosystem and am curious about the reason for the change and if it is needed to go through the Manage private apps steps or there's a better way for a project wiht an existing accessToken.

Thanks Much!
